### DISCUSSION TITLE
Information regarding transpose solver in custom_linear_solve

### DISCUSSION THREAD

**Discussion by SNMS95** [jax-ml/jax#16757] [Q&A] [answered]
Hey guys,

I am trying to use a PETSc linear solver alongside JAX.
I have a residual function r(u, rho)=0, which I solve in the forward pass to get to the system of equations Ax = b, which is solved with this solver.
In order to have a better control of the entire process, I implemented a `custom_jvp` rule for the forward pass. Since I want to support higher-order autodiff for the forward pass, I understand that the `custom_jvp` rule has to be completely traceable by JAX. 

That is when I saw `jax.lax.custom_linear_solve`, which can help me specify the rule since it does not care whether the solver used is blackbox or not. I use the PETSc solver as the `solve` attribute, with the correct signature i.e. `solve=petsc_solver(matvec, rhs)`.

However, the `transpose_solve` option is something that I couldn't understand well. The documentation says that it should be a function that can solve vecmat instead of matvec. 
1. What should be it's signature?
2. Can I use the same petsc_solver function for this ?

The psuedo-code I use is follows:
```
#jax.custom_jvp
def forward_pass(params):
    solution = complex_non_jax_function(params)  # This should be a Primitive but for now, I am interested only in Auto-diff 
    return solution

@forward_solve.defjvp
def forward_pass_jvp(primals, tangents):
    params, = primals
    params_dot, = tangents
    primal_out = forward_pass(params)  # This should be traceable
    # For calculating the JVP rule
    # 1. Define residual function
    def residual(solution, params): # r(u, p) = 0 
          .....
         return value
    # 2. Construct terms for JVP rule
    # 2.1 dr/du and dr/dp - using jax.linearize and jax.jvp respectively
    # They form the matvec and rhs respectively
    
    # 3. Solve adjoint to get du/dp
     tangents_out = jax.lax.custom_linear_solve(matvec, rhs,
                                                                                   solve= petsc_solver, transpose_solve = ???????????)

    return tangents_out
    

```

**Comment by patrick-kidger** [comment]
If you're looking to create a custom linear solver, I'd recommend trying [Lineax](https://github.com/google/lineax) instead of `custom_linear_solve`. This is our newly-released more-sophisticated approach to linear solvers. In particular this will handle the details of transposition for you. You can create a custom solver by subclassing `lineax.AbstractLinearSolver`.

(To explain why this argument is there, though: the transpose solve is needed in order to handle backpropagation, as this involves solving the linear system `A^T x = b`.)

**Reply by SNMS95** [reply]
Thanks Patrick, 

So, you are suggesting that I should subclass the PETSc solver ? and then use it normally in the jvp rule?

**Reply by SNMS95** [reply]
i.e. the only thing I would need to change is to use the subclassed PETSc solver in Step 3. to calculate the adjoints right?

**Reply by patrick-kidger** [reply]
I'm suggesting something like:
```python
class PetscSolver(lineax.AbstractLinearSolver):
    ...  # fill in the details of calling your solver

solver = PetscSolver()
soluion = lineax.linear_solve(A, b, solver)
```
The details of calculating adjoints will be handled automatically; you just need to describe how the forward pass works.

**Reply by SNMS95** [reply]
Is there a documentation for creating such a sub-class?

**Reply by patrick-kidger** [reply]
Yes, some: see the API definition of `AbstractLinearSolver` [here](https://docs.kidger.site/lineax/api/solvers/#lineax.AbstractLinearSolver), which specifies all the methods you need to implement.

The source code for an existing solvers will also a useful reference; [here's the source code for an LU solver](https://github.com/google/lineax/blob/main/lineax/_solver/lu.py), which is pretty simple.

**Comment by SNMS95** [comment]
I understand that I can use `Lineax` but since I am short on time, I would like to make it work with `jax.lax.custom_linear_solver`.
It works well when I use a jax solver as the `solve`. However, as soon as I change it to black-box solver, this throws a `TraceError`.

I thought that the solver internals did not matter for this function but it seems that solver is traced as well!

**Reply by jakevdp** [reply, marked as answer]
Yes, I believe the solver system will be traced here. If your goal is to call to an external solver from JAX code, I think `custom_linear_solve` is not the solution you're looking for. Perhaps [`jax.pure_callback`]() will do what you have in mind?

See [External Callbacks in JAX](https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html) for more information on this.
